File: articles/overfitting-underfitting-and-model-generalization/fortran/generalization_summary.f90
program generalization_summary
  implicit none

  integer, parameter :: n = 5
  integer :: i, best_index
  character(len=32) :: model_id(n)
  character(len=32) :: class_label
  real(8) :: training_rmse(n), validation_rmse(n), complexity(n), interpretability(n)
  integer :: parameter_count(n)
  real(8) :: current_score, best_score, gap

  model_id = (/ 'constant_baseline               ', 'linear_trend                    ', 'logistic_growth                 ', 'regularized_curve               ', 'high_flex_curve                 ' /)
  training_rmse = (/3.40d0, 1.95d0, 1.20d0, 0.95d0, 0.28d0/)
  validation_rmse = (/3.55d0, 2.10d0, 1.38d0, 1.44d0, 2.85d0/)
  parameter_count = (/0, 2, 3, 5, 10/)
  complexity = (/0.05d0, 0.25d0, 0.45d0, 0.62d0, 0.95d0/)
  interpretability = (/0.95d0, 0.88d0, 0.78d0, 0.66d0, 0.30d0/)

  best_score = huge(1.0d0)
  best_index = 1

  print '(A)', 'model_id generalization_score overfit_gap classification'

  do i = 1, n
    current_score = validation_rmse(i) + 0.20d0 * complexity(i) &
      + 0.08d0 * real(parameter_count(i), 8) - 0.20d0 * interpretability(i)
    gap = validation_rmse(i) - training_rmse(i)

    if (training_rmse(i) >= 3.0d0 .and. validation_rmse(i) >= 3.0d0) then
      class_label = 'likely_underfit'
    else if (gap >= 1.0d0 .and. training_rmse(i) <= 1.0d0) then
      class_label = 'likely_overfit'
    else if (validation_rmse(i) <= 1.5d0 .and. gap <= 0.6d0) then
      class_label = 'generalizes_reasonably'
    else
      class_label = 'requires_review'
    end if

    print '(A,1X,F8.4,1X,F8.4,1X,A)', trim(model_id(i)), current_score, gap, trim(class_label)
    if (current_score < best_score) then
      best_score = current_score
      best_index = i
    end if
  end do

  print '(A,1X,A)', 'selected_for_review', trim(model_id(best_index))

end program generalization_summary

File: articles/overfitting-underfitting-and-model-generalization/cpp/generalization_summary.cpp
#include <iomanip>
#include <iostream>
#include <limits>
#include <string>
#include <vector>

struct Candidate {
    std::string id;
    std::string family;
    double training_rmse;
    double validation_rmse;
    int parameter_count;
    double complexity;
    double interpretability;
};

double score(const Candidate& model) {
    return model.validation_rmse
        + 0.20 * model.complexity
        + 0.08 * static_cast<double>(model.parameter_count)
        - 0.20 * model.interpretability;
}

std::string classify(const Candidate& model) {
    const double gap = model.validation_rmse - model.training_rmse;
    if (model.training_rmse >= 3.0 && model.validation_rmse >= 3.0) {
        return "likely_underfit";
    }
    if (gap >= 1.0 && model.training_rmse <= 1.0) {
        return "likely_overfit";
    }
    if (model.validation_rmse <= 1.5 && gap <= 0.6) {
        return "generalizes_reasonably";
    }
    return "requires_review";
}

int main() {
    std::vector<Candidate> models = {
        {"constant_baseline", "baseline", 3.40, 3.55, 0, 0.05, 0.95},
        {"linear_trend", "statistical", 1.95, 2.10, 2, 0.25, 0.88},
        {"logistic_growth", "mechanistic", 1.20, 1.38, 3, 0.45, 0.78},
        {"regularized_curve", "regularized", 0.95, 1.44, 5, 0.62, 0.66},
        {"high_flex_curve", "flexible", 0.28, 2.85, 10, 0.95, 0.30}
    };

    double best_score = std::numeric_limits<double>::infinity();
    std::string best_id;

    std::cout << "model_id,generalization_score,overfit_gap,classification\n";
    for (const auto& model : models) {
        const double current_score = score(model);
        const double gap = model.validation_rmse - model.training_rmse;
        std::cout << std::fixed << std::setprecision(6)
                  << model.id << "," << current_score << "," << gap << "," << classify(model) << "\n";
        if (current_score < best_score) {
            best_score = current_score;
            best_id = model.id;
        }
    }

    std::cerr << "selected_for_review=" << best_id << " generalization_score=" << best_score << "\n";
    return 0;
}
